torch/csrc/api/include/torch/data/iterator.h

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+#include <torch/csrc/utils/variadic.h>
+#include <torch/tensor.h>
+
+#include <ATen/core/Error.h>
+
+#include <functional>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace torch {
+namespace data {
+namespace detail {
+// For increased safety and more separated logic, this implementation of
+// `Iterator` consists of a `ValidIterator` and a `SentinelIterator`. A
+// `ValidIterator` yields new batches until the `DataLoader` is exhausted. While
+// the `DataLoader` is not exhausted, `ValidIterator`s compare equal if they are
+// the same object. When the `ValidIterator` becomes exhauted, it compares equal
+// to the `SentinelIterator`, but not before. Half the code here is to implement
+// double dispatch for the comparison. Got damnit, C++.
+
+template <typename Batch>
+struct ValidIterator;
+
+template <typename Batch>
+struct SentinelIterator;
+
+/// Base class for the `ValidIterator` and `SentinelIterator`
+template <typename Batch>
+struct IteratorImpl {
+  virtual ~IteratorImpl() = default;
+  virtual void next() = 0;
+  virtual Batch& get() = 0;
+  virtual bool operator==(const IteratorImpl& other) const = 0;
+  virtual bool operator==(const ValidIterator<Batch>& other) const = 0;
+  virtual bool operator==(const SentinelIterator<Batch>& other) const = 0;
+};
+
+template <typename Batch>
+struct ValidIterator : public IteratorImpl<Batch> {
+  using BatchProducer = std::function<optional<Batch>()>;
+
+  explicit ValidIterator(BatchProducer next_batch)
+      : next_batch_(std::move(next_batch)) {}
+
+  /// Fetches the next batch.
+  void next() override {
+    // If we didn't get the very first batch yet, get it now.
+    lazy_initialize();
```

Comment:
I don't understand. This can trigger initialization, which gets the first batch, that you discard immediately afterwards?!

Reply:
The contract with `Iterator` is that when you call `auto iter = begin()`, then `iter` either points to the first valid item, or equals the end. `next()` fetches the next item. The lazy initialization is so that only the first dereference actually fetches the item (if you start using `iter` a couple lines after creating it). If I write:

```cpp
auto iter = begin();
iter.next();
```

Then the user should not know about the lazy initialization of course, and since `iter` first pointed at the first item, the call to `next()` should make it point at the second item. Therefore we first need to (lazily) fetch the first item, and then increment to the second item. Hope this makes sense.